### PR TITLE
feat: add localization provider

### PR DESCRIPTION
**Pull request by Korisss**
Add LocalizationProvider.

### CODE REVIEW

**Comment by Korisss** [comment]
😂
image

**Comment by Korisss** [comment]
review plz

File: skymp5-functions-lib/src/types/mp.d.ts [hunk 1]
```diff
@@ -113,7 +113,7 @@ export interface Mp {
   get(formId: 0, propertyName: 'onlinePlayers'): number[];
   get(formId: number, propertyName: 'isDead'): boolean;
   get(formId: number, propertyName: 'worldOrCellDesc'): string;
-  get(formId: number, propertyName: 'percentages'): { health: number, magicka: number, stamina: number };
+  get(formId: number, propertyName: 'percentages'): { health: number; magicka: number; stamina: number };
```
Comment (Wedmer) [inline skymp5-functions-lib/src/types/mp.d.ts:116]: Was this change necessary?

File: skymp5-functions-lib/src/types/mp.d.ts [hunk 2]
```diff
@@ -126,7 +126,11 @@ export interface Mp {
   set(formId: number, propertyName: 'locationalData', newValue: LocationalData): void;
   set(formId: number, propertyName: 'spawnPoint', newValue: LocationalData): void;
   set(formId: number, propertyName: 'isDead', newValue: boolean): void;
-  set(formId: number, propertyName: 'percentages', newValue: { health: number, magicka: number, stamina: number }): void;
+  set(
```
Comment (Wedmer) [inline skymp5-functions-lib/src/types/mp.d.ts:129]: And this?

File: skymp5-server/cpp/addon/main.cc [hunk 3]
```diff
@@ -316,6 +320,9 @@ ScampServer::ScampServer(const Napi::CallbackInfo& info)
     const auto& logger = GetLogger();
     partOne->AttachLogger(logger);
 
+    logger->info("Run localization provider");
+    this->localizationProvider = new LocalizationProvider("russian");
```
Comment (Wedmer) [inline skymp5-server/cpp/addon/main.cc:324]: Maybe this thing should be configured via config file?

**Review by Wedmer — CHANGES_REQUESTED** [review]
Almost lgtm. Just answer questions.

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+std::map<uint32_t, std::string> LocalizationProvider::ParseStrings(
+  std::vector<char> buffer, std::vector<DirectoryEntry*> entries)
+{
+  std::map<uint32_t, std::string> localization;
+  uint32_t numberOfEntries = entries.size();
+
+  for (int i = 0; i < numberOfEntries; i++) {
+    uint32_t start = 8 + numberOfEntries * 8 + (*entries[i]).offset;
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:41]:
```suggestion
    uint32_t start = 8 + numberOfEntries * 8 + entries[i]->offset;
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    for (int charIndex = 0; charIndex < buffer.size() - start; charIndex++) {
+      if (buffer[start + charIndex] == 0) {
+        break;
+      }
+
+      (*entries[i]).str += buffer[start + charIndex];
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:48]: Same, `->`

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    localization[(*entries[i]).stringId] = (*entries[i]).str;
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:51]: Same, `->`

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+std::map<uint32_t, std::string> LocalizationProvider::ParseILDLStrings(
+  std::vector<char> buffer, std::vector<DirectoryEntry*> entries)
+{
+  std::map<uint32_t, std::string> localization;
+  uint32_t numberOfEntries = entries.size();
+
+  for (int i = 0; i < numberOfEntries; i++) {
+    uint32_t start = 8 + numberOfEntries * 8 + (*entries[i]).offset;
+
+    (*entries[i]).length = *(uint32_t*)&(buffer[start]);
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:66]: `->` and C++ cast

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    for (int charIndex = 0; charIndex < (*entries[i]).length - start;
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:68]: `->`

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+         charIndex++) {
+      if (buffer[start + 4 + charIndex] == 0) {
+        break;
+      }
+
+      (*entries[i]).str += buffer[start + 4 + charIndex];
+    }
+
+    localization[(*entries[i]).stringId] = (*entries[i]).str;
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:77]: same

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    std::vector<char> buffer, std::vector<DirectoryEntry*> entries);
+
+  std::map<uint32_t, std::string> Parse(
+    const std::filesystem::directory_entry& file);
+
+public:
+  LocalizationProvider(std::string language);
+  std::map<std::string, std::map<uint32_t, std::string>>
+    localization; // localization[filename][stringId]
+  std::string Get(std::string file, uint32_t stringId);
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.h:35]:
```suggestion
  std::string Get(const std::string& file, uint32_t stringId);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    std::vector<char> buffer, std::vector<DirectoryEntry*> entries);
+
+  std::map<uint32_t, std::string> ParseILDLStrings(
+    std::vector<char> buffer, std::vector<DirectoryEntry*> entries);
+
+  std::map<uint32_t, std::string> Parse(
+    const std::filesystem::directory_entry& file);
+
+public:
+  LocalizationProvider(std::string language);
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.h:32]:
```suggestion
  LocalizationProvider(const std::string& language);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  std::map<std::string, std::map<uint32_t, std::string>>
+    localization; // localization[filename][stringId]
```
Comment (nic11) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.h:34]: Does this need to be publicly accessible?

File: skymp5-functions-lib/src/types/mp.d.ts (hunk 1 above)
Comment (Korisss) [reply skymp5-functions-lib/src/types/mp.d.ts:116]: Oops, I have automatic formatting in VS Code with prettier

File: skymp5-server/cpp/addon/main.cc (hunk 3 above)
Comment (Korisss) [reply skymp5-server/cpp/addon/main.cc:324]: Thanks, I forgot

File: skymp5-functions-lib/src/types/mp.d.ts [hunk 13]
```diff
@@ -126,7 +126,7 @@ export interface Mp {
   set(formId: number, propertyName: 'locationalData', newValue: LocationalData): void;
   set(formId: number, propertyName: 'spawnPoint', newValue: LocationalData): void;
   set(formId: number, propertyName: 'isDead', newValue: boolean): void;
-  set(formId: number, propertyName: 'percentages', newValue: { health: number, magicka: number, stamina: number }): void;
+  set(formId: number, propertyName: 'percentages', newValue: { health: number; magicka: number; stamina: number } ): void;
```
Comment (Wedmer) [inline skymp5-functions-lib/src/types/mp.d.ts:129]: Formatting changes are still here. We don't need them until we will reformat all scripts with some codestyle.

**Review by Wedmer — CHANGES_REQUESTED** [review]

File: skymp5-server/cpp/addon/main.cc [hunk 14]
```diff
@@ -966,6 +973,54 @@ void ScampServer::RegisterChakraApi(std::shared_ptr<JsEngine> chakraEngine)
...
+      if (lookupRes.rec) {
+        auto fields = JsValue::Array(0);
+
+        auto& cache = partOne->worldState.GetEspmCache();
+
+        espm::IterateFields_(
+          lookupRes.rec,
+          [&](const char* type, uint32_t size, const char* data) {
+            if (std::string(type, 4) == "FULL" && size == 4) {
+              auto stringId = *(uint32_t*)data;
```
[the remark this answers is not in the document]
Comment (Wedmer) [reply skymp5-server/cpp/addon/main.cc:993]:
```suggestion
              auto stringId = *reinterpret_cast<uint32_t*>(data);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+  uint32_t length;
+  std::string str;
+
+  DirectoryEntry();
+  DirectoryEntry(uint32_t stringId, uint32_t stringOffset);
+};
+
+class LocalizationProvider
+{
+  std::vector<DirectoryEntry*> ParseDirectoryEntries(std::vector<char> buffer);
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.h:20]: Bad design. Raw pointers used such way are road to memory leaks.

```suggestion
  std::vector<shared_ptr<DirectoryEntry> > ParseDirectoryEntries(const std::vector<char> &buffer);
```

```suggestion
  void ParseDirectoryEntries(const std::vector<char> &buffer, std::vector<DirectoryEntry> &container);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    std::string filename = entry.path().filename().string();
+
+    if (!entry.is_directory() &&
+        filename.find(language) != std::string::npos) {
+      size_t lastIndex = filename.find_last_of("_");
+      std::string name = filename.substr(0, lastIndex);
+
+      auto parsedStrings = this->Parse(entry);
+
+      this->localization[name].insert(parsedStrings.begin(),
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:124]: `this->` is not needed.

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+  for (const auto& entry : std::filesystem::directory_iterator(stringsPath)) {
+    std::string filename = entry.path().filename().string();
+
+    if (!entry.is_directory() &&
+        filename.find(language) != std::string::npos) {
+      size_t lastIndex = filename.find_last_of("_");
+      std::string name = filename.substr(0, lastIndex);
+
+      auto parsedStrings = this->Parse(entry);
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:122]: You can achieve less copies by passing name to `Parse` and filling `LocalizationProvider::localization` directly there.
`this->` is not needed.

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    this->ParseDirectoryEntries(buffer);
+
+  if (file.path().extension() == ".strings") {
+    return this->ParseStrings(buffer, directoryEntries);
+  } else if (file.path().extension() == ".dlstrings" ||
+             file.path().extension() == ".ilstrings") {
+    return this->ParseILDLStrings(buffer, directoryEntries);
+  }
+
+  throw std::runtime_error("Unexcepted file in ./data/strings/ path: " +
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:106]: Throwing exceptions in function used inside constructor without catches is not a good idea.

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+std::map<uint32_t, std::string> LocalizationProvider::Parse(
+  const std::filesystem::directory_entry& file)
+{
+  std::vector<char> buffer;
+  std::ifstream fileStream(file.path(), std::ios::binary);
+
+  buffer.resize(file.file_size());
+
+  if (!fileStream.read(buffer.data(), file.file_size())) {
+    throw std::runtime_error("[Localization Provider] can't read " +
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:92]: Throwing exceptions in function used inside constructor without catches is not a good idea.

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+    directoryEntries[i] = entry;
+  }
+
+  return directoryEntries;
+}
+
+std::map<uint32_t, std::string> LocalizationProvider::ParseStrings(
+  std::vector<char> buffer, std::vector<DirectoryEntry*> entries)
+{
+  std::map<uint32_t, std::string> localization;
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:37]: Bad thing to use same name for class field and variable in class function. Also you can fill class field directly from this function.

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+    localization[(*entries[i]).stringId] = (*entries[i]).str;
+  }
+
+  return localization;
+}
+
+std::map<uint32_t, std::string> LocalizationProvider::ParseILDLStrings(
+  std::vector<char> buffer, std::vector<DirectoryEntry*> entries)
+{
+  std::map<uint32_t, std::string> localization;
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:60]: Bad thing to use same name for class field and variable in class function. Also you can fill class field directly from this function.

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+
+  std::map<uint32_t, std::string> ParseStrings(
+    std::vector<char> buffer, std::vector<DirectoryEntry*> entries);
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.h:23]: Don't copy heavy containers. Pass them by a reference. Use `const` for data that shouldn't be modified. 
```suggestion
    const std::vector<char> &buffer, std::vector<DirectoryEntry*> &entries);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+  std::map<uint32_t, std::string> ParseILDLStrings(
+    std::vector<char> buffer, std::vector<DirectoryEntry*> entries);
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.h:26]: Don't copy heavy containers. Pass them by a reference. Use const for data that shouldn't be modified.
```suggestion
    const std::vector<char> &buffer, std::vector<DirectoryEntry*> &entries);
```

**Review by Wedmer — CHANGES_REQUESTED** [review]
After further reading I've found that this class should be little redesigned.

File: skymp5-server/cpp/addon/main.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (Korisss) [reply skymp5-server/cpp/addon/main.cc:993]: After this fix I get error C2440 `invalid conversion from "const char *" to "uint32_t *" `

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h (hunk 23 above)
Comment (Korisss) [reply skymp5-server/cpp/localization_provider/LocalizationProvider.h:26]: if i use "const" i get C2440 error when trying to convert buffer to uint32_t

File: skymp5-server/cpp/addon/main.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (Wedmer) [reply skymp5-server/cpp/addon/main.cc:993]: > After this fix I get error C2440 `invalid conversion from "const char *" to "uint32_t *" `

```suggestion
              auto stringId = *reinterpret_cast<const uint32_t*>(data);
```

File: skymp5-functions-lib/src/types/mp.d.ts [hunk 24]
```diff
@@ -126,7 +126,7 @@ export interface Mp {
   set(formId: number, propertyName: 'locationalData', newValue: LocationalData): void;
   set(formId: number, propertyName: 'spawnPoint', newValue: LocationalData): void;
   set(formId: number, propertyName: 'isDead', newValue: boolean): void;
-  set(formId: number, propertyName: 'percentages', newValue: { health: number, magicka: number, stamina: number }): void;
+  set(formId: number, propertyName: 'percentages', newValue: { health: number, magicka: number, stamina: number } ): void;
```
Comment (Wedmer) [inline skymp5-functions-lib/src/types/mp.d.ts:129]: Still changed with a space character.

File: skymp5-server/cpp/addon/main.cc [hunk 25]
```diff
@@ -117,6 +119,8 @@ class ScampServer : public Napi::ObjectWrap<ScampServer>
   Viet::TaskQueue chakraTaskQueue;
   GamemodeApi::State gamemodeApiState;
 
+  LocalizationProvider* localizationProvider;
```
Comment (Wedmer) [inline skymp5-server/cpp/addon/main.cc:122]: It's better to use shared pointer + make_shared on initialization.

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h (hunk 23 above)
Comment (Wedmer) [reply skymp5-server/cpp/localization_provider/LocalizationProvider.h:26]: If data shouldn't be changed use `reinterpret_cast<const type>(variable);`

File: skymp5-server/cpp/addon/main.cc [hunk 26]
```diff
@@ -360,6 +364,16 @@ ScampServer::ScampServer(const Napi::CallbackInfo& info)
...
 
+    if (serverSettings["lang"] == nullptr) {
+      serverSettings["lang"] = "english";
+    }
+
+    logger->info("Run localization provider for language:",
+                 serverSettings["lang"]);
+    this->localizationProvider =
+      std::shared_ptr<LocalizationProvider>(new LocalizationProvider(
+        serverSettings["dataDir"], serverSettings["lang"]));
```
Comment (Wedmer) [inline skymp5-server/cpp/addon/main.cc:375]: It's better to use `std::make_shared` to write less.
```suggestion
    localizationProvider = std::make_shared<LocalizationProvider>(serverSettings["dataDir"], serverSettings["lang"]);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+  DirectoryEntry(uint32_t stringId, uint32_t stringOffset);
+};
+
+class LocalizationProvider
+{
+  std::map<std::string, std::map<uint32_t, std::string>>
+    localization; // localization[filename][stringId]
+
+  std::vector<DirectoryEntry> ParseDirectoryEntries(
+    const std::vector<char>& buffer);
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.h:24]:
```suggestion
   void ParseDirectoryEntries(
    const std::vector<char>& buffer, std::vector<DirectoryEntry>& entries);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+#include <fstream>
+
+DirectoryEntry::DirectoryEntry(uint32_t stringId, uint32_t stringOffset)
+  : stringId(stringId)
+  , offset(stringOffset)
+{
+}
+
+std::vector<DirectoryEntry> LocalizationProvider::ParseDirectoryEntries(
+  const std::vector<char>& buffer)
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:11]:
```suggestion
void LocalizationProvider::ParseDirectoryEntries(
  const std::vector<char>& buffer, std::vector<DirectoryEntry>& entries)
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+{
+  uint32_t numberOfEntries = *reinterpret_cast<const uint32_t*>(&buffer[0]);
+
+  std::vector<DirectoryEntry> entries;
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:15]:
```suggestion
  entries.clear();
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+
+    uint32_t stringOffset =
+      *reinterpret_cast<const uint32_t*>(&buffer[offset]);
+    offset += 4;
+
+    DirectoryEntry entry(stringId, stringOffset);
+    entries[i] = entry;
+  }
+
+  return entries;
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:32]:
```suggestion
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 31]
```diff
@@ -0,0 +1,130 @@
...
+  std::ifstream fileStream(file.path(), std::ios::binary);
+
+  buffer.resize(file.file_size());
+
+  if (!fileStream.read(buffer.data(), file.file_size())) {
+    throw std::runtime_error("[Localization Provider] can't read " +
+                             file.path().string());
+  }
+
+  std::vector<DirectoryEntry> directoryEntries = ParseDirectoryEntries(buffer);
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:96]:
```suggestion
  std::vector<DirectoryEntry> directoryEntries;  
  ParseDirectoryEntries(buffer, directoryEntries);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.h [hunk 32]
```diff
@@ -0,0 +1,38 @@
...
+
+  void ParseILDLStrings(std::string name, const std::vector<char>& buffer,
+                        std::vector<DirectoryEntry>& entries);
+
+  void Parse(const std::filesystem::directory_entry& file);
+
+public:
+  LocalizationProvider(const std::string& dataDir,
+                       const std::string& language);
+  std::string Get(const std::string& file, uint32_t stringId);
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.h:37]:
```suggestion
  const std::string& Get(const std::string& file, uint32_t stringId);
```

File: skymp5-server/cpp/localization_provider/LocalizationProvider.cpp [hunk 33]
```diff
@@ -0,0 +1,130 @@
...
+
+    if (!entry.is_directory() &&
+        filename.find(language) != std::string::npos) {
+
+      Parse(entry);
+    }
+  }
+}
+
+std::string LocalizationProvider::Get(const std::string& file,
```
Comment (Wedmer) [inline skymp5-server/cpp/localization_provider/LocalizationProvider.cpp:126]:
```suggestion
const std::string& LocalizationProvider::Get(const std::string& file,
```

**Review by Wedmer — CHANGES_REQUESTED** [review]
Showed places where you can reduce unnecessary copies.

File: skymp5-server/cpp/addon/main.cc [hunk 34]
```diff
@@ -360,6 +364,14 @@ ScampServer::ScampServer(const Napi::CallbackInfo& info)
...
     }
 
+    if (serverSettings["lang"] == nullptr) {
+      serverSettings["lang"] = "english";
+    }
+
+    logger->info("Run localization provider");
+    localizationProvider = std::make_shared<LocalizationProvider>(
+      serverSettings["dataDir"], serverSettings["lang"]);
+
```
Comment (Wedmer) [inline skymp5-server/cpp/addon/main.cc:374]: I think that we should handle variant when this functionality is not needed. 
Maybe it's better to detect which languages we have onboard, so we will need some option like "useLocalizationProvider" with boolean value.
For now just make your new code not terminating initialization process if there're no strings files.
Something like:
```suggestion
    if (serverSettings["lang"] != nullptr) {
      logger->info("Run localization provider");
      localizationProvider = std::make_shared<LocalizationProvider>(
        serverSettings["dataDir"], serverSettings["lang"]);
    }
```

File: skymp5-server/cpp/addon/main.cc [hunk 35]
```diff
@@ -966,6 +978,54 @@ void ScampServer::RegisterChakraApi(std::shared_ptr<JsEngine> chakraEngine)
...
+                        fileNameWithoutExt, stringId));
+                  }
+                }
+              }
+            }
+          },
+          cache);
+      }
+
+      return translatedString;
```
Comment (Wedmer) [inline skymp5-server/cpp/addon/main.cc:1026]: If `localizationProvider` wasn't initialized we'll need a guard.
```suggestion
      auto translatedString = JsValue::Undefined();
      if(localizationProvider){
        auto globalRecordId = ExtractFormId(args[1], "globalRecordId");
        auto lookupRes =
          partOne->GetEspm().GetBrowser().LookupById(globalRecordId);
        
        if (lookupRes.rec) {
          auto fields = JsValue::Array(0);

          auto& cache = partOne->worldState.GetEspmCache();

          espm::IterateFields_(
            lookupRes.rec,
            [&](const char* type, uint32_t size, const char* data) {
              if (std::string(type, 4) == "FULL" && size == 4) {
                auto stringId = *reinterpret_cast<const uint32_t*>(data);
                if (serverSettings["loadOrder"].is_array()) {
                  for (size_t i = 0; i < serverSettings["loadOrder"].size();
                       ++i) {
                    std::filesystem::path loadOrderElement =
                      static_cast<std::string>(serverSettings["loadOrder"][i]);

                    if (i == lookupRes.fileIdx) {
                      auto fileNameFull = loadOrderElement.filename().string();
                      auto fileNameWithoutExt =
                        fileNameFull.substr(0, fileNameFull.find_last_of("."));

                      std::transform(
                        fileNameWithoutExt.begin(), fileNameWithoutExt.end(),
                        fileNameWithoutExt.begin(),
                        [](unsigned char c) { return std::tolower(c); });

                      translatedString =
                        JsValue::String(this->localizationProvider->Get(
                          fileNameWithoutExt, stringId));
                    }
                  }
                }
              }
            },
            cache);
        }
      }
      return translatedString;
```

**Review by Wedmer — CHANGES_REQUESTED** [review]
So it's finally looks ok. Only two edits needed for merge. Also there are some thoughts about making this part better.

**Review by Wedmer — APPROVED** [review]
